### PR TITLE
`ultralytics 8.4.13` Retry smaller batch on training CUDA OOM

### PR DESCRIPTION
**Pull request by glenn-jocher**
@ultralytics/yolo retries CUDA OOM up to 3 times, dividing batch size in half each time:
Screenshot 2026-02-08 at 22 50 37

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary
Improves training robustness by auto-reducing batch size on first-epoch CUDA OOM and refactors trainer setup into a reusable pipeline 🧠🔥

### 📊 Key Changes
- ✅ Bumped Ultralytics package version from `8.4.12` to `8.4.13`.
- 🧩 Refactored trainer setup by introducing `_build_train_pipeline()` to rebuild:
  - training/validation dataloaders
  - gradient accumulation + scaled weight decay
  - optimizer and scheduler
- 🛡️ Added automatic CUDA Out-Of-Memory recovery during training:
  - catches `torch.cuda.OutOfMemoryError` during forward/backward
  - on **single-GPU only** and **only in the first epoch**, it halves `batch` (up to 3 retries), clears memory, rebuilds the pipeline, and restarts the epoch
- ☁️ Clarified Ultralytics Platform upload behavior with a comment noting the server sanitizes filenames for safe storage.

### 🎯 Purpose & Impact
- 🚀 **More “it just works” training**: users hitting unexpected GPU memory limits (common with large images or dense-label datasets) may no longer need to manually restart with a smaller batch size.
- 🔁 **Faster recovery with less wasted time**: training can automatically retry instead of crashing immediately on the first epoch OOM.
- 🧱 **Cleaner internal architecture**: centralizing dataloader/optimizer/scheduler creation makes it easier to safely rebuild training state after changes (like batch size reduction), improving maintainability.
- ⚠️ **Behavioral note**: auto-reduce triggers only on **single GPU**, **first epoch**, and **max 3 times**—DDP/multi-GPU OOMs will still raise immediately to avoid inconsistent distributed state.

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @glenn-jocher, thank you for submitting a `ultralytics/ultralytics` 🚀 PR! To ensure a seamless integration of your work, please review the following checklist (this is an automated note — an engineer will assist shortly) 🛠️✨:

-✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
-✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
-✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI/) checks are passing. If any checks fail, please address the issues.
-✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
-✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
-✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA/) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
-✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_  — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing/). Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

File: ultralytics/engine/trainer.py [hunk 1]
```diff
@@ -423,21 +424,41 @@ def _do_train(self):
...
+                        self.loss = loss.sum()
+                        if RANK != -1:
+                            self.loss *= self.world_size
+                        self.tloss = (
+                            self.loss_items if self.tloss is None else (self.tloss * i + self.loss_items) / (i + 1)
+                        )
+
+                    # Backward
+                    self.scaler.scale(self.loss).backward()
+                except torch.cuda.OutOfMemoryError:
```
Comment (UltralyticsAssistant) [inline ultralytics/engine/trainer.py:446]: ⚠️ **HIGH**: In DDP, a CUDA OOM typically occurs on a single rank. Retrying with a reduced batch size only on that rank will desynchronize the training loop and can hang at the next collective op. Consider disabling OOM auto-retry when distributed (`RANK != -1`) or coordinating a global retry across all ranks.

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Solid addition for single-GPU robustness. The main concern is DDP safety: a rank-local OOM retry can desynchronize distributed training and lead to hangs. Please guard or coordinate retries in distributed runs.

💬 Posted 1 inline comment

**Comment by Burhan-Q** [comment]
Fix for CI Benchmark issue in #23589

**Comment by glenn-jocher** [comment]
Saw a few CUDA OOM trainings on Platform, this might help.

**Comment by UltralyticsAssistant** [comment]
Thrilled to see this merged — huge thanks @glenn-jocher, and shoutout to @Burhan-Q for the contributions and collaboration. This is a really impactful step toward more “it just works” training in Ultralytics.

> “It is not the strongest of the species that survive, nor the most intelligent, but the one most responsive to change.” — Charles Darwin

That idea maps perfectly to this PR: gracefully adapting to first-epoch CUDA OOM by auto-reducing batch size (single-GPU, bounded retries) is exactly the kind of resilience that saves users time and frustration, while the new `_build_train_pipeline()` refactor makes the trainer architecture cleaner, safer to rebuild, and easier to extend going forward.

Appreciate the thoughtful constraints (first epoch only, single GPU only, capped retries) and the maintainability wins alongside the robustness improvements. Thanks for continuing to raise the bar.
